docs/proposals/vip-4.md

```diff
@@ -0,0 +1,273 @@
+  a delay between each region swap. The code change to support these extra timestamps is included in this VIP, to 
+  demonstrate feasibility and present the proposed algorithm (see `OfflinePushStatus::getStatusUpdateTimestamp`).
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
+
+1. Within the push job.
+2. Within the parent controller.
+3. Within the child controllers.
+
+It is also possible to consider invoking some hooks in one of these location while other hooks would be executed 
+elsewhere. The table below summarizes the feasibility and tradeoffs for each of the proposed hooks:
+
+|                           Hook function name |  Actions   | CC  | PC  | VPJ |
+|---------------------------------------------:|:----------:|:---:|:---:|:---:|
+|                         `validateHookConfig` |   ➡️ ☠️    |  ❌  |  ✅  |  ❌  |
+|                          `preStartOfPushJob` |   ➡️ ☠️    |  ❌  |  ✅  |  ✅  |
+|                         `postStartOfPushJob` |    None    |  ❌  |  ✅  |  ✅  |
+|                      `preSchemaRegistration` |   ➡️ ☠️    |  ❌  |  ✅  | 1️⃣ |
+|                     `postSchemaRegistration` |    None    |  ✅  |  ✅  | 1️⃣ |
+|                    `preStoreVersionCreation` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                   `postStoreVersionCreation` |    None    |  ✅  |  ✅  |  ✅  |
+|  `preStartOfStoreVersionIngestionForDaVinci` | ➡️ ☠️ ✋ ↩️ | 2️⃣ | 2️⃣ | 2️⃣ |
+| `postStartOfStoreVersionIngestionForDaVinci` | ➡️ ☠️ ✋ ↩️ | 2️⃣ | 2️⃣ | 2️⃣ |
+|          `postStoreVersionLeaderReplication` |    None    |  ✅  |  ❌  |  ❌  |
+|                        `preStoreVersionSwap` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                       `postStoreVersionSwap` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                            `preEndOfPushJob` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                           `postEndOfPushJob` |    None    |  ✅  |  ✅  |  ✅  |
+
+**Legend:**
+
+* The Actions column represent which control mechanism is available to each hook:
+  * ➡️ Proceed: Move forward with this step.
+  * ☠️ Abort: Cancel this step (and as a consequence, short-circuit any future step that would come after this one).
+  * ✋ Wait: Let the hooks framework re-run this step later (i.e. 1 minute later, by default).
+  * ↩️ Rollback: Let the store go back to the store-version it had prior to beginning the push job (in all regions). 
+
+* The last three columns are the feasibility of implementing this hook in a given component:
+  * **CC** Child Controller.
+  * **PC** Parent Controller.
+  * **VPJ** Venice Push Job.
+  * ✅ It is feasible to implement this hook within this component (without unreasonable complexity).
+  * ❌ It is NOT feasible to implement this hook within this component (without unreasonable complexity).
+  * 1️⃣ Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and 
```

Comment:
Just curious, wouldn't schema registration have the same consequence regardless of the trigger?

Reply:
Yes, the consequences should be the same regardless of the trigger. The point of this line is to say that if we execute the schema registration hooks inside VPJ, then we will have a blindspot by design (which is not ideal...).

Reply:
Got it.

And `Schema registration` hook will be invoked regardless of the trigger, right?

Reply:
Only if the hook is executed in the controller...